File: 5Nodes/MultiParameterSweeper.m
% Set the path of 64 bit NetSim Binaries to be used for simulation.
NETSIM_PATH = "C:\Program Files\NetSim\Standard_v14_3\bin\bin_x64";

LICENSE_ARG = "5053@192.168.0.4"; %Floating license
%LICENSE_ARG = """C:\Program Files\NetSim\Standard_v13_2\bin"""; %Node Locked license

% Set NETSIM_AUTO environment variable to avoid keyboard interrupt at the end of each simulation
setenv('NETSIM_AUTO','1');

% Create IOPath directory to store the input Configuration.netsim file and
% the simulation output files during each iteration
if(~isfolder('IOPath'))
    mkdir IOPath
end

% Create Data directory to store the Configuration.netsim and the
% Metrics.xml files associated with each iteration
if(~isfolder('Data'))
    mkdir Data
end

% Clear the IOPath folder if it has any files created during previous
delete('IOPath\*')

% Delete result.csv file if it already exists
if(isfile('result.csv'))
    delete result.csv
end

% create a csv file to log the output metrics for analysis csvfile
fileid=fopen('result.csv','w');

% Add headings to the CSV file
fprintf(fileid,'X1,X2,X3,X4,X5,Range,Packets Received');
fclose(fileid);

rng('default')
ran_arr = 1000*rand(100,5);
ran_arr = sort(ran_arr,2);

% create a folder with name as year-month-day-hour.minute.seconds inside the data folder
today = datestr(now,'dd-mm-yyyy-HH.MM.SS');
mkdir('Data',today)

% Iterate based on the number of time simulation needs to be run and the
for j = 50:50:1000

    for i = 1:100
    
        if(isfile('Configuration.netsim'))
            delete Configuration.netsim
        end
        
        if(isfile("IOPath\\Configuration.netsim"))
            delete('IOPath\Configuration.netsim')
        end
        
        if(isfile('IOPath\Metrics.xml'))
            delete('IOPath\Metrics.xml')
        end
        
        %Call ConfigWriter.exe with arguments as per the number of variable parameters in the input.xml file
        cmd="ConfigWriter.exe " + string(ran_arr(i,1)) + " "  + string(ran_arr(i,2)) + " " + string(ran_arr(i,3)) + " "  + string(ran_arr(i,4))+ " "+ string(ran_arr(i,5)) + " " +  string(j);    
        disp(cmd)
        system(cmd);
        
        %Copy the Configuration.netsim file generated by ConfigWriter.exe to IOPath directory
        if(isfile('Configuration.netsim'))
            copyfile('Configuration.netsim','IOPath');
            copyfile('ConfigSupport\*', 'IOPath\ConfigSupport');
        end
        
        strIOPATH=pwd+"\IOPath";
        
        %Run NetSim via CLI mode by passing the apppath iopath and license information to the NetSimCore.exe
        cmd="start ""NetSim_Multi_Parameter_Sweeper"" /wait /d "...
        +""""... 
        +NETSIM_PATH...
        + """ "... 
        + "NetSimcore.exe -apppath """...
        +NETSIM_PATH...
        + """ -iopath """... 
        +strIOPATH... 
        +""" -license "...
        +LICENSE_ARG;
            
        %disp(cmd)
        system(cmd);
        
        %Create a copy of the output Metrics.xml file for writing the result log
        if(isfile('IOPath\Metrics.xml'))
            copyfile('IOPath\Metrics.xml','Metrics.xml');
            system("MetricsCsv.exe IOPath");
        end    
      
        %If only one output parameter is to be read only one Script text file with name Script.txt to be provided
        %If more than one output parameter is to be read, multiple Script text file with name Script1.txt, Script2.txt,...
        %...,Scriptn.txt to be provided
        
        OUTPUT_PARAM_COUNT=1;
        
        %changes reqiured starts
        if(isfile('IOPath\Metrics.xml'))
            
            fileid=fopen('result.csv','a');
            if(fileid)
                fprintf(fileid,'\n%f,%f,%f,%f,%f,%f,',ran_arr(i,1),ran_arr(i,2),ran_arr(i,3),ran_arr(i,4),ran_arr(i,5),j);
                fclose(fileid);
            end
            if(OUTPUT_PARAM_COUNT==1)
                system("MetricsReader.exe result.csv");
            else
                for n = 1 : OUTPUT_PARAM_COUNT
                    filename='Script'+string(n)+'.txt';
                    movefile(filename,'Script.txt');
                    system("MetricsReader.exe result.csv");
                    fileid=fopen('result.csv','a');
                    if(fileid)
                        fprintf(fileID,',');
                        fclose(fileid);
                    end
                    movefile('Script.txt',filename);             
                end
            end
        else
            %Update the output Metric as crash if Metrics.xml file is missing
            fileid=fopen('result.csv','a');
            if(fileid)
                fprintf(fileid,'\n%f,%f,%f,%f,%f,%f,crash',ran_arr(i,1),ran_arr(i,2),ran_arr(i,3),ran_arr(i,4),ran_arr(i,5),j);
                fclose(fileid);
            end
        end
    
        % Name of the Output folder to which the result will be saved
        OUTPUT_PATH = "Data\"+today+"\Range_"+string(j)+"\Output_"+string(i);
        
        if(~isfolder(OUTPUT_PATH))
            mkdir(OUTPUT_PATH)
        end
        
        % Create a copy of result.csv file present in sweep folder to date-time
        if(isfile('result.csv'))
            copyfile('result.csv', "Data\"+today)
        end
        
        % Create a copy of all files that is present in IOPATH to the desired output location
        movefile('IOPath\*',OUTPUT_PATH)
        
        % Delete Configuration.netsim file created during the last iteration
        if(isfile("Configuration.netsim"))
            delete('Configuration.netsim')
        end
        
        if(isfile('Metrics.xml'))
            delete('Metrics.xml')
        end
    end
end

File: 10Nodes/MultiParameterSweeper.py
import subprocess
import random
import shutil
import numpy as np
import time
import os

# Set the path of 64-bit NetSim Binaries to be used for simulation.
NETSIM_PATH = "C:\\Users\\MIC\\Documents\\NetSim\\Workspaces\\Defv14-3-6\\bin_x64"
LICENSE_ARG = "5053@192.168.0.4"

# Set NETSIM_AUTO environment variable to avoid keyboard interrupt at the end of each simulation
os.environ["NETSIM_AUTO"] = "1"

# Create directories if they don’t exist
os.makedirs("IOPath", exist_ok=True)
os.makedirs("Data", exist_ok=True)

# Clear IOPath directory from previous runs
for root, dirs, files in os.walk("IOPath"):
    for file in files:
        os.remove(os.path.join(root, file))

# Delete result.csv if it exists
if os.path.isfile("result.csv"):
    os.remove("result.csv")

# Create a CSV file to log output metrics
with open("result.csv", "w") as csvfile:
    csvfile.write("X1,X2,X3,X4,X5,X6,X7,X8,X9,X10,Range,Packets Received,")

# Create random UE positions
np.random.seed(0)
ran_arr = np.round(1000 * np.random.rand(100, 10), 2)
ran_arr = np.sort(ran_arr, axis=1)

# Create a timestamped output folder
foldername = time.strftime("%Y-%m-%d-%H.%M.%S")

# Iterate through parameter range
for j in range(50, 1001, 50):
    for i in range(100):  # Fixed indentation issue here

        # Remove old files before running the simulation
        for filename in ["Configuration.netsim", "IOPath\\Configuration.netsim", "IOPath\\Metrics.xml"]:
            if os.path.isfile(filename):
                os.remove(filename)

        # Call ConfigWriter.exe
        cmd = f'ConfigWriter.exe {ran_arr[i,0]} {ran_arr[i,1]} {ran_arr[i,2]} {ran_arr[i,3]} {ran_arr[i,4]} {ran_arr[i,5]} {ran_arr[i,6]} {ran_arr[i,7]} {ran_arr[i,8]} {ran_arr[i,9]} {j}'
        print(cmd)
        os.system(cmd)

        # Copy generated Configuration.netsim file
        if os.path.isfile("Configuration.netsim"):
            shutil.copy("Configuration.netsim", "IOPath\\Configuration.netsim")

        # Copy Additional Files
        if os.path.exists('ConfigSupport'):
            shutil.rmtree("IOPath\\ConfigSupport", ignore_errors=True)
            shutil.copytree("ConfigSupport", "IOPath\\ConfigSupport")

        # Run NetSim via CLI
        strIOPATH = os.getcwd() + "\\IOPath"
        cmd = (
            f'start "NetSim_Multi_Parameter_Sweeper" /wait /d "{NETSIM_PATH}" '
            f'NetSimcore.exe -apppath "{NETSIM_PATH}" -iopath "{strIOPATH}" -license {LICENSE_ARG}'
        )
        os.system(cmd)

        # Process Metrics.xml if available
        OUTPUT_PARAM_COUNT = 1
        if os.path.isfile("IOPath\\Metrics.xml"):
            shutil.copy("IOPath\\Metrics.xml", "Metrics.xml")
            os.system("MetricsCsv.exe IOPath")

            with open("result.csv", "a") as csvfile:
                csvfile.write(f"\n{ran_arr[i,0]},{ran_arr[i,1]},{ran_arr[i,2]},{ran_arr[i,3]},{ran_arr[i,4]},{ran_arr[i,5]},{ran_arr[i,6]},{ran_arr[i,7]},{ran_arr[i,8]},{ran_arr[i,9]},{j},")

            if OUTPUT_PARAM_COUNT == 1:
                subprocess.run(["MetricsReader.exe", "result.csv"], check=True)
            else:
                for n in range(1, OUTPUT_PARAM_COUNT + 1):
                    os.rename(f"Script{n}.txt", "Script.txt")
                    subprocess.run(["MetricsReader.exe", "result.csv"], check=True)
                    with open("result.csv", "a") as csvfile:
                        csvfile.write(",")
                    os.rename("Script.txt", f"Script{n}.txt")

        else:
            # If Metrics.xml is missing, mark it as "crash"
            with open("result.csv", "a") as csvfile:
                csvfile.write(f"\n{i},crash,crash,crash,crash,crash,crash,")

        # Save output results
        OUTPUT_PATH = f"Data\\{foldername}\\Range_{j}\\Output_{i}"
        #OUTPUT_PATH = f"Data\\{foldername}\\Output_{i}"
        os.makedirs(OUTPUT_PATH, exist_ok=True)

        if os.path.isfile("result.csv"):
            shutil.copy("result.csv", f"Data\\{foldername}")

        for file_name in os.listdir("IOPath"):
            shutil.move(os.path.join("IOPath", file_name), OUTPUT_PATH)

        # Safe deletion with retry for locked Metrics.xml
        for filename in ["Configuration.netsim", "Metrics.xml"]:
            if os.path.isfile(filename):
                for _ in range(5):  # Retry up to 5 times
                    try:
                        os.remove(filename)
                        break
                    except PermissionError:
                        print(f"File {filename} is locked, retrying in 2 seconds...")
                        time.sleep(2)
